rskj-core/src/main/java/co/rsk/net/AsyncNodeBlockProcessor.java

```diff
@@ -0,0 +1,227 @@
+        boolean readyForProcessing = blockSyncService.preprocessBlock(block, sender, false);
+        if (readyForProcessing) {
+            // Validate block if it can be added to the queue for processing
+            if (isBlockValid(block)) {
+                boolean offer = blocksToProcess.offer(new BlockInfo(sender, block));
+                if (offer) {
+                    logger.trace("Added block with number {} and hash {} from {} to the queue", blockNumber, blockHash, peer);
+                } else {
+                    // This should not happen as the queue is unbounded
+                    logger.warn("Cannot add block for processing into the queue with number {} {} from {}", blockNumber, blockHash, peer);
+                }
+
+                return scheduledForProcessingResult(block, start);
+            }
+
+            logger.warn("Invalid block with number {} {} from {} ", blockNumber, blockHash, peer);
+            return invalidBlockResult(block, start);
+        }
+
+        logger.trace("Ignored block with number {} and hash {} from {} as it's not ready for processing yet", blockNumber, blockHash, peer);
+        return ignoreBlockResult(block, start);
+    }
+
+    @Override
+    public void start() {
+        thread.start();
+    }
+
+    @Override
+    public void stop() {
+        stopThread();
+    }
+
+    /**
+     * Stop the service and wait until a working thread is stopped for {@code waitMillis} milliseconds,
+     * if {@code waitMillis} greater than zero. If {@code waitMillis} is zero, then immediately returns.
+     */
+    public void stopAndWait(long waitMillis) throws InterruptedException {
+        stopThread();
+
+        if (waitMillis > 0L) {
+            thread.join(waitMillis);
+        }
+    }
+
+    private void stopThread() {
+        stopped = true;
+        thread.interrupt();
+    }
+
+    @Override
+    public void run() {
+        while (!stopped) {
+            Peer sender = null;
+            Block block = null;
+
+            try {
+                logger.trace("Awaiting block for processing from the queue...");
+
+                BlockInfo blockInfo = blocksToProcess.take();
+
+                sender = blockInfo.peer;
+                block = blockInfo.block;
+
+                if (logger.isTraceEnabled()) {
+                    logger.trace("Start block processing with number {} and hash {} from {}", block.getNumber(), block.getPrintableHash(), sender);
+                }
+
+                BlockProcessResult blockProcessResult = blockSyncService.processBlock(block, sender, false);
+                logger.trace("Finished block processing");
+
+                if (listener != null) {
+                    listener.onBlockProcessed(this, sender, block, blockProcessResult);
+                }
+            } catch (InterruptedException e) {
+                logger.trace("Thread has been interrupted");
+
+                Thread.currentThread().interrupt();
```

Comment:
stopped = true? Could be an undesired stop, right?

Reply:
well, this flag `stopped` is being set by other threads. This particular thread with the loop only reads it.